scripts/rv_expression_association/get_vep_variants.py

```diff
@@ -0,0 +1,94 @@
+
+import logging
+import hail as hl
+import numpy as np
+import pandas as pd
+from hail.methods import export_plink
+from cpg_utils.hail_batch import dataset_path, init_batch, output_path
+
+# object containing variants within a 50K window on either side of the VPREB3 gene
+MT = dataset_path('v0/VPREB3_50K_window.mt')
+VEP_HT = dataset_path('v0/VPREB3_50K_window_vep.ht')
+
+# use logging to print statements, display at info level
+logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)
+
+
+def main():
+    # read and densify object
+    init_batch()
+    mt = hl.read_matrix_table(MT)
+    logging.info(f'Number of variants in window: {mt.count()[0]}')
+
+    mt = hl.experimental.densify(mt)
+    # filter out low quality variants and consider biallelic variants only (no multi-allelic, no ref-only)
+    mt = mt.filter_rows(
+        (hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)
+        & (hl.len(mt.alleles) == 2)
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))
+    )
+
+    # annotate using VEP
+    vep_ht = hl.read_table(VEP_HT)
+    mt = mt.annotate_rows(vep=vep_ht[mt.row_key].vep)
+    mt_path = output_path('densified_qced_snps_only_vep_annotated.mt', 'tmp')
+    mt = mt.checkpoint(
+        mt_path, overwrite=True
+    )  # add checkpoint to avoid repeat evaluation
+    logging.info(f'Number of QC-passing, biallelic SNPs: {mt.count()[0]}')
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    rv_mt = mt.filter_rows((mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0))
+    rv_mt_path = output_path('rare_variants.mt', 'tmp')
+    rv_mt = rv_mt.checkpoint(rv_mt_path, overwrite=True)
+    logging.info(f'Number of rare variants (freq<5%): {rv_mt.count()[0]}')
+
+    # filter variants found to have regulatory effects
+    filtered_mt = rv_mt.filter_rows(
+        hl.len(rv_mt.vep.regulatory_feature_consequences['biotype']) > 0
+    )
+    logging.info(
+        f'Number of rare variants (freq<5%) with ergulatory conequences: {filtered_mt.count()[0]}'
+    )
+
+    filtered_rrv_mt = filtered_mt.filter_rows(filtered_mt.variant_qc.AF[1] < 0.01)
+    logging.info(f'Number of rarer variants (freq<1%): {filtered_rrv_mt.count()[0]}')
+
+    filtered_0maf_mt = filtered_mt.filter_rows(filtered_mt.variant_qc.AF[1] == 0)
+    logging.info(
+        f'Check that there are {filtered_0maf_mt.count()[0]} variants with freq==0'
+    )
+
+    # print out stats
+    # types of regulatory variants
+    biotypes = pd.Series(
+        filtered_mt.vep.regulatory_feature_consequences['biotype'].collect()
+    )
+    print(
+        biotypes.value_counts()
+    )  # not sure how to turn this into logging? Multiline table
+
+    # MAF distribution
+    mafs = pd.Series(filtered_mt.variant_qc.AF[1].collect())
+    logging.info(f'Min frequency in set: {np.nanmin(mafs)}')
+    logging.info(f'Max frequency in set: {np.nanmax(mafs)}')
+    logging.info(f'Mean frequency in set: {np.nanmean(mafs)}')
```

Comment:
`collect` doesn't scale well, as it'll require a lot of memory on a single node. Could you use Hail's `stats` method here instead?
https://hail.is/docs/0.2/aggregators.html#hail.expr.aggregators.stats

Reply:
great, will definitely use Hail's `stats` method to get min, max and mean! However if I wanted to save info in a table like i mentioned before, i was going to use to `collect` to get maf values, add them to a pandas data frame and then save as csv? Are you suggesting to use a Hail `table` instead?

Reply:
follow up, i tried to save as a hail table, maybe check syntax? then, i can delete commented out lines

Reply:
Looks like you're now saving the full table (I've added comments there, too).

But coming back to your previous question: you can easily serialize the `struct` that https://hail.is/docs/0.2/aggregators.html#hail.expr.aggregators.stats returns, e.g. to JSON, in case you need to save that for downstream analyses.

Reply:
I think if I save the whole table these stats should be easy to pull out when i need them in the future? E.g. I might select variants that vep annotated to be promoter, and check the allele frequency distribution for those variants specifically, etc?

Reply:
Yup, makes sense!